src/algebra/category/Module/change_of_rings.lean

```diff
@@ -161,4 +166,166 @@ variables {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →
+  { intros, rw [ring_hom.id_apply, restrict_scalars.smul_def, ←mul_smul, mul_comm, mul_smul,
+      restrict_scalars.smul_def], },
+  { intros, ext, simp only [linear_map.add_apply, linear_map.coe_mk, add_smul], },
+  { intros, ext,
+    simpa only [ring_hom.id_apply, linear_map.smul_apply, linear_map.coe_mk,
+      @restrict_scalars.smul_def _ _ _ _ f ⟨S⟩, smul_eq_mul, mul_smul], },
+  { intros, rw [map_add], },
+  { intros s z,
+    rw [ring_hom.id_apply],
+    induction z using tensor_product.induction_on with x s' z1 z2 ih1 ih2,
+    { simp only [smul_zero, map_zero], },
+    { simp only [extend_scalars.smul_tmul, linear_map.coe_mk, tensor_product.lift.tmul, mul_smul] },
+    { rw [smul_add, map_add, map_add, ih1, ih2, smul_add], } },
+end
+
+/--
+The natural transformation from the composition of restriction and extension of scalars to the
+identity functor on `S`-module.
+-/
+@[simps] def counit : (restrict_scalars f ⋙ extend_scalars f) ⟶ (𝟭 (Module S)) :=
+{ app := λ _, counit.map f,
+  naturality' := λ Y Y' g,
+    begin
+      ext z, induction z using tensor_product.induction_on,
+      { simp only [map_zero] },
+      { simp only [category_theory.functor.comp_map, Module.coe_comp, function.comp_app,
+          extend_scalars.map_tmul, restrict_scalars.map_apply, counit.map_apply, lift.tmul,
+          linear_map.coe_mk, category_theory.functor.id_map, linear_map.map_smulₛₗ,
+          ring_hom.id_apply] },
+      { simp only [map_add, *] },
+    end }
+
+end extend_restrict_scalars_adj
+
+/--
+Given commutative rings `R, S` and a ring hom `f : R →+* S`, the extension and restriction of
+scalars by `f` are adjoint to each other.
+-/
+def extend_restrict_scalars_adj {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S]
+  (f : R →+* S) : extend_scalars f ⊣ restrict_scalars f :=
+{ hom_equiv := λ _ _, extend_restrict_scalars_adj.hom_equiv f,
+  unit := extend_restrict_scalars_adj.unit f,
+  counit := extend_restrict_scalars_adj.counit f,
+  hom_equiv_unit' := λ X Y g, linear_map.ext $ λ x, by simp,
+  hom_equiv_counit' := λ X Y g, linear_map.ext $ λ x,
+    begin
+      induction x using tensor_product.induction_on,
+      { simp only [map_zero]},
+      { simp only [extend_restrict_scalars_adj.hom_equiv_symm_apply, linear_map.coe_mk,
+        extend_restrict_scalars_adj.hom_equiv.from_extend_scalars_apply, tensor_product.lift.tmul,
+        extend_restrict_scalars_adj.counit_app, Module.coe_comp, function.comp_app,
+        extend_scalars.map_tmul, extend_restrict_scalars_adj.counit.map_apply] },
+      { simp only [map_add, *], }
+    end }
+
+instance {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →+* S) :
+  category_theory.is_left_adjoint (extend_scalars f) := ⟨_, extend_restrict_scalars_adj f⟩
+
+instance {R : Type u₁} {S : Type u₂} [comm_ring R] [comm_ring S] (f : R →+* S) :
+  category_theory.is_right_adjoint (restrict_scalars f) := ⟨_, extend_restrict_scalars_adj f⟩
+
```

Comment:
Since `restrict_scalars` is fully faithful, the counit is a natural isomorphism.
I think this result should be stated as well.

Reply:
I agree that `restrict_scalars` is faithful, but I don't think it is full. When $f :  \mathbb{Z}\to S$, `restrict_scalars` is the forgetful functor from module to abelian groups which is not full

Reply:
Yeah, if N is an S-module then the counit sends S tensor_R N to N and this is sometimes an isomorphism (e.g. if S is a localisation of R) but not always (e.g. if R=reals and C=complexes).